### PR TITLE
noise: Initial version of the noise chat plugin

### PR DESCRIPTION
**Pull request by cdecker**
Based on the WhatSat idea by @joostjager, this plugin implements a simple chat
protocol based on top of `createonion` and `sendonion`.

It still needs a bit of cleanup and needs some pending changes to c-lightning, but as soon as they are in I'll undraft this :wink: 

Depends: https://github.com/ElementsProject/lightning/pull/3261 https://github.com/ElementsProject/lightning/pull/3260 and needs to allow TLV payloads

### CODE REVIEW

File: noise/noise.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    payload.write(msg.encode('UTF-8'))
+
+    # Sign the message:
+    sig = plugin.rpc.signmessage(msg)['signature']
+    sig = unhexlify(sig)
+    varint_encode(34349336, payload)
+    varint_encode(len(sig), payload)
+    payload.write(sig)
+
+    res = deliver(node_id, payload.getbuffer(), amt=amt)
```
Comment (joostjager) [inline noise/noise.py]: No key send to compensate the routing nodes for their service?

Comment (cdecker) [reply noise/noise.py]: I have a local branch that does it, but I needed to have a working branch for the `createonion` and `sendonion` pull request. It's really just a 2 line change to add payments attached to messages :wink:

File: noise/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+zbase32==1.1.5
+pyln-client>=0.0.7.3
```
Comment (darosior) [inline noise/requirements.txt]:
```suggestion
pyln-client>=0.7.3
```

File: noise/noise.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        "payload": hexlify(payload).decode('ASCII'),
+    })
+    return first_hop, hops, route
+
+
+def deliver(node_id, payload, amt, max_attempts=5, payment_hash=None):
+    """Do your best to deliver `payload` to `node_id`.
+    """
+    if payment_hash is None:
+        payment_hash = ''.join(random.choice(string.hexdigits) for _ in range(64)).lower()
```
Comment (darosior) [inline noise/noise.py]: nit
```suggestion
        payment_hash = ''.join(random.choices(string.hexdigits, k=64)).lower()
```

**Comment by cdecker** [comment]
> Can be undrafted now that [ElementsProject/lightning#3261](https://github.com/ElementsProject/lightning/pull/3261) [ElementsProject/lightning#3260](https://github.com/ElementsProject/lightning/pull/3260) were merged? :-)

Absolutely, I'd like to clean things up a bit first, but we should be able to merge soon ;-)

**Comment by cdecker** [comment]
Ok, had to work around a few quirks (`zbase32` encoded signature from `signmessage`, better TLV encoding support), but this should not work pretty nicely:

 - The sender is recovered from the message (TLV type had to change so we don't collide with WhatSat)
 - A payment can now be sent alongside a message, using the `keysend` protocol (using a different TLV type again, since c-lightning does not allow types above `0xFFFFFFFF`, I'll fix that asap)

This should be good to review @darosior and @renepickhardt  :-)

File: noise/README.org [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+* Protocol
+The protocol was heavily inspired by the [[https://github.com/joostjager/whatsat#protocol][WhatSat protocol]]:
+
+| record type | length (bytes) | value                                                           |
+|-------------+----------------+-----------------------------------------------------------------|
+|  5482373484 |             32 | key send preimage                                               |
+|    34349334 |       variable | chat message                                                    |
+|    34349335 |             65 | compressed signature + recovery id                              |
+|    34349339 |             33 | sender pubkey                                                   |
+|    34349343 |              8 | timestamp in nano seconds since unix epoch (big endian encoded) |
```
Comment (darosior) [inline noise/README.org]: Why does the first record type has a value that higher from the four following types ?

File: noise/noise.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+
+def deliver(node_id, payload, amt, payment_hash, max_attempts=5):
+    """Do your best to deliver `payload` to `node_id`.
+    """
+    exclusions = []
+    payment_hash = hexlify(payment_hash).decode('ASCII')
+
+    for attempt in range(max_attempts):
+        logging.debug("Starting attempt {} to deliver message to {}".format(attempt, node_id))
```
Comment (darosior) [inline noise/noise.py]: Is this another logging trick ? :)

File: noise/requirements.txt (hunk 2 above)
Comment (darosior) [reply noise/requirements.txt]: Reviving this comment too

File: noise/onion.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
+from primitives import varint_decode, varint_encode
```
Comment (darosior) [inline noise/onion.py]: Do you plan to include this (`noise/onion.py`) to pyln-proto?

File: noise/primitives.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import struct
```
Comment (darosior) [inline noise/primitives.py]: This too ?

File: noise/onion.py (hunk 6 above)
Comment (darosior) [reply noise/onion.py]: I should have checked C-lightning side..

**Comment by darosior** [comment]
Twitter-tested ACK 0bbaa0c2508eac9924279a7802fbbe840b26ecb6 (https://twitter.com/darosior/status/1219642537188741123?s=20) :grin:

**Comment by rsbondi** [comment]
This seems to be one behind on the receiving.  If I do 
`l1-cli sendmsg ID3 "message 0"`
`l3-cli recvmsg 0` is waiting, and does not retrieve message 0

if I send another message from `l1-cli` and am now at id `1` the `recvmsg 0` call to finally returns but with message id 1, 

I can then call `l3-cli recvmsg 0` and get message 0, but `l3-cli recvmsg 1` is now waiting and not retrieving the received message until the next message is sent.  I can repeat with the same results. so if I am receiving on id 10 I can call `recvmsg` with id 0-9 but can not get id 10.

The readme file says "You can read the last message received" with `recvmsg` but that is not what I am seeing, I am seeing one behind.

It seems to work if the receiver is expecting the next message with `recvmsg -1`(seems to work with any number <= next expected id) and it will wait and pull in messages as they come, but I am sure this will timeout at some point

**Comment by darosior** [comment]
Got a crash because of a missing method for `LegacyOnionPayload`, used in response to `htlc_accepted` (encounetered while I was rebalancing two channels).
```
**BROKEN** lightningd: Plugin for htlc_accepted returned non-result response {"error": "Error while processing htlc_accepted: AttributeError(\"'LegacyOnionPayload' object has no attribute 'get'\",)", "id": 82, "jsonrpc": "2.0"}
```
```

File: noise/noise.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+@plugin.hook('htlc_accepted')
+def on_htlc_accepted(onion, htlc, plugin, **kwargs):
+    payload = OnionPayload.from_hex(onion['payload'])
+
+    msg = Message(
+        id=len(plugin.messages),
+        sender=None,
+        body=payload.get(34349334).value,
+        signature=payload.get(34349335).value,
+        payment=None)
```
Comment (rsbondi) [inline noise/noise.py]: this is always null

File: noise/README.org (hunk 4 above)
Comment (cdecker) [reply noise/README.org]: The choice was done by @joostjager, but if we ever standardize this we'd be chosing smaller types, for the sake of conserving space in the onion.

**Comment by cdecker** [comment]
> This seems to be one behind on the receiving. If I do
> `l1-cli sendmsg ID3 "message 0"`
> `l3-cli recvmsg 0` is waiting, and does not retrieve message 0
> 
> if I send another message from `l1-cli` and am now at id `1` the `recvmsg 0` call to finally returns but with message id 1,
> I can then call `l3-cli recvmsg 0` and get message 0, but `l3-cli recvmsg 1` is now waiting and not retrieving the received message until the next message is sent. I can repeat with the same results. so if I am receiving on id 10 I can call `recvmsg` with id 0-9 but can not get id 10.
> 
> The readme file says "You can read the last message received" with `recvmsg` but that is not what I am seeing, I am seeing one behind.
> 
> It seems to work if the receiver is expecting the next message with `recvmsg -1`(seems to work with any number <= next expected id) and it will wait and pull in messages as they come, but I am sure this will timeout at some point

This is working as expected, though the documentation might be sub-par: the argument to `recvmsg` is the last message you processed, so we don't assume a monotonically increasing message ID, but rather let the plugin assign arbitrary IDs, and the parameter acts almost as an ACK from the caller to the plugin that the previous message was processed correctly.

The reason why I went this way is that it matches the RPCs we have in c-lightning, and it makes it easier once we have different topics that we might want to tail. If we have several chat sessions open with different users for example we might want the next message (after receiving message `i`) from Alice, but the next incoming message (`i+1`) might be from Bob, so asking for `i+1` would get us an incorrect message from a completely different chat session. With the high-water mark approach it makes a lot more sense to ask for "Messages from Alice since the message with ID `i`".

How could I formulate this in the README.md to be clearer?

File: noise/noise.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    signature_field = payload.get(34349335)
+
+    if body_field is None or signature_field is None:
+        plugin.log("Missing message body or signature, ignoring HTLC")
+        return {'result': 'continue'}
+
+    msg = Message(
+        id=len(plugin.messages),
+        sender=None,
+        body=body_field.value,
```
Comment (m-schmoock) [inline noise/noise.py]: Hey, as we did the mainchain test couple of days ago, I had a crash in the log in this or the next line when accessing unchecked attribute `value` on `body_field` or ` signature_field` I guess:

```
2020-01-22T21:59:38.842Z **BROKEN** lightningd: Plugin for htlc_accepted returned non-result response {"jsonrpc": "2.0", "id": 48, "error": "Error while processing htlc_accepted: AttributeError(\"'NoneType' object has no attribute 'value'\")"}
2020-01-22T21:59:38.986Z **BROKEN** lightningd: FATAL SIGNAL 6 (version v0.8.0-modded)
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: common/daemon.c:46 (send_backtrace) 0x556c67af1d69
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: common/daemon.c:54 (crashdump) 0x556c67af1db9
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe067adfaf
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe067adf25
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe06797896
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: lightningd/log.c:819 (fatal) 0x556c67ac573c
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: lightningd/plugin_hook.c:83 (plugin_hook_callback) 0x556c67ae68e8
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: lightningd/plugin.c:258 (plugin_response_handle) 0x556c67ae2b3f
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/plugin.c:356 (plugin_read_json_one) 0x556c67ae2ce7
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/plugin.c:388 (plugin_read_json) 0x556c67ae2e3a
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x556c67b43445
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/io.c:407 (do_plan) 0x556c67b43fc6
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/io.c:417 (io_ready) 0x556c67b44004
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/poll.c:445 (io_loop) 0x556c67b461d0
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/io_loop_with_timers.c:24 (io_loop_with_timers) 0x556c67abbfc6
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/lightningd.c:855 (main) 0x556c67ac2d2f
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe06799152
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x556c67aa73cd
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0xffffffffffffffff
```

File: noise/README.org [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+* Protocol
+The protocol was heavily inspired by the [[https://github.com/joostjager/whatsat#protocol][WhatSat protocol]]:
+
+| record type | length (bytes) | value                                                           |
+|-------------+----------------+-----------------------------------------------------------------|
+|  5482373484 |             32 | key send preimage                                               |
+|    34349334 |       variable | chat message                                                    |
+|    34349335 |             65 | compressed signature + recovery id                              |
+|    34349339 |             33 | sender pubkey                                                   |
```
Comment (rsbondi) [inline noise/README.org]: this is not used, pubkey is derived from the signature

File: noise/noise.py (hunk 9 above)
Comment (cdecker) [reply noise/noise.py]: Yep, fixed that in the last round of fixes, we now check that the necessary fields are passed in ok, otherwise we fail the HTLC.

**Pull request merged**